pkg/openstack/instancesv2.go

```diff
@@ -0,0 +1,193 @@
+	if !os.useV1Instances {
+		return os.instancesv2()
+	}
+	return nil, false
+}
+
+func (os *OpenStack) instancesv2() (*InstancesV2, bool) {
+	klog.V(4).Info("openstack.Instancesv2() called")
+
+	compute, err := client.NewComputeV2(os.provider, os.epOpts)
+	if err != nil {
+		klog.Errorf("unable to access compute v2 API : %v", err)
+		return nil, false
+	}
+
+	regionalProviderID := false
+	if isRegionalProviderID := sysos.Getenv(RegionalProviderIDEnv); isRegionalProviderID == "true" {
+		regionalProviderID = true
+	}
+
+	return &InstancesV2{
+		compute:          compute,
+		region:           os.epOpts.Region,
+		regionProviderID: regionalProviderID,
+		networkingOpts:   os.networkingOpts,
+	}, true
+}
+
+// InstanceExists indicates whether a given node exists according to the cloud provider
+func (i *InstancesV2) InstanceExists(ctx context.Context, node *v1.Node) (bool, error) {
+	_, err := i.getInstance(ctx, node)
+	if err == cloudprovider.InstanceNotFound {
+		klog.V(6).Infof("instance not found for node: %s", node.Name)
+		return false, nil
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// InstanceShutdown returns true if the instance is shutdown according to the cloud provider.
+func (i *InstancesV2) InstanceShutdown(ctx context.Context, node *v1.Node) (bool, error) {
+	server, err := i.getInstance(ctx, node)
+	if err != nil {
+		return false, err
+	}
+
+	// SHUTOFF is the only state where we can detach volumes immediately
+	if server.Status == instanceShutoff {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+type MetadataCache struct {
+	Name     string
+	Metadata *cloudprovider.InstanceMetadata
+}
+
+// InstanceMetadata returns the instance's metadata.
+func (i *InstancesV2) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
+	srv, err := i.getInstance(ctx, node)
+	if err != nil {
+		return nil, err
+	}
+	server := ServerAttributesExt{}
+	if srv != nil {
+		server = *srv
+	}
+
+	instanceType, err := srvInstanceType(i.compute, &server.Server)
+	if err != nil {
+		return nil, err
+	}
+
+	interfaces, err := getAttachedInterfacesByID(i.compute, server.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	addresses, err := nodeAddresses(&server.Server, interfaces, i.networkingOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cloudprovider.InstanceMetadata{
+		ProviderID:    i.makeInstanceID(&server.Server),
+		InstanceType:  instanceType,
+		NodeAddresses: addresses,
+		Zone:          server.AvailabilityZone,
+		Region:        i.region,
+	}, nil
+}
+
+func (i *InstancesV2) makeInstanceID(srv *servers.Server) string {
+	if i.regionProviderID {
+		return fmt.Sprintf("%s://%s/%s", ProviderName, i.region, srv.ID)
+	}
+	return fmt.Sprintf("%s:///%s", ProviderName, srv.ID)
+}
+
+func (i *InstancesV2) getInstance(ctx context.Context, node *v1.Node) (*ServerAttributesExt, error) {
+	if node.Spec.ProviderID == "" {
+		opt := servers.ListOpts{
+			Name: node.Name,
```

Comment:
Bizarrely this is a regular expression sub-match. Assuming we want an exact match here I think this needs to be:

```go
Name: fmt.Sprintf("^%s$", node.Name)
```

Reference: https://docs.openstack.org/api-ref/compute/?expanded=list-servers-detail#list-servers (see `name` request parameter)